File: apps/web/src/store/ui-store.ts
'use client';

import { create } from 'zustand';
import type { ThemeName } from '../lib/themes';

interface UIStore {
  theme: ThemeName;
  isDragging: boolean;
  dragPiece: {
    piece: string;
    fromRow: number;
    fromCol: number;
  } | null;
  boardFlipped: boolean;
  showNewGameDialog: boolean;
  mobileHistoryOpen: boolean;
  // Replay mode for offline game
  replayMode: boolean;
  replayStep: number;

  setTheme: (theme: ThemeName) => void;
  startDrag: (piece: string, fromRow: number, fromCol: number) => void;
  endDrag: () => void;
  flipBoard: () => void;
  setShowNewGameDialog: (show: boolean) => void;
  toggleMobileHistory: () => void;
  enterReplay: (step: number) => void;
  exitReplay: () => void;
  setReplayStep: (step: number) => void;
}

export const useUIStore = create<UIStore>((set) => ({
  theme: 'classic',
  isDragging: false,
  dragPiece: null,
  boardFlipped: true,
  showNewGameDialog: false,
  mobileHistoryOpen: false,
  replayMode: false,
  replayStep: 0,

  setTheme: (theme) => set({ theme }),

  startDrag: (piece, fromRow, fromCol) =>
    set({ isDragging: true, dragPiece: { piece, fromRow, fromCol } }),

  endDrag: () =>
    set({ isDragging: false, dragPiece: null }),

  flipBoard: () =>
    set((state) => ({ boardFlipped: !state.boardFlipped })),

  setShowNewGameDialog: (show) =>
    set({ showNewGameDialog: show }),

  toggleMobileHistory: () =>
    set((state) => ({ mobileHistoryOpen: !state.mobileHistoryOpen })),

  enterReplay: (step) => set({ replayMode: true, replayStep: step }),
  exitReplay: () => set({ replayMode: false, replayStep: 0 }),
  setReplayStep: (step) => set({ replayStep: step }),
}));

File: apps/Ios-app/Raichu/Raichu/Stores/UIStore.swift
// UIStore.swift
// Raichu
// UI preferences: theme, board orientation, replay mode.

import SwiftUI
import Combine

@MainActor
final class UIStore: ObservableObject {
    @Published var themeName: String = "classic" {
        didSet {
            UserDefaults.standard.set(themeName, forKey: "raichuTheme")
        }
    }
    @Published var boardFlipped: Bool = true {  // white at bottom by default
        didSet {
            UserDefaults.standard.set(boardFlipped, forKey: "raichuBoardFlipped")
        }
    }
    /// Selected root tab. Lives here so screens like AuthView can navigate
    /// without threading a binding through the view tree.
    @Published var selectedTab: Int = Tab.play
    @Published var replayMode: Bool = false
    @Published var replayStep: Int = 0
    @Published var hapticsEnabled: Bool = true {
        didSet {
            UserDefaults.standard.set(hapticsEnabled, forKey: "raichuHaptics")
            HapticManager.shared.isEnabled = hapticsEnabled
        }
    }

    var currentTheme: ThemeConfig {
        ThemeConfig.theme(named: themeName)
    }

    init() {
        // `object(forKey:) as? Bool` rather than `bool(forKey:)`: the latter
        // returns false for an unset key, which would invert both defaults.
        themeName = UserDefaults.standard.string(forKey: "raichuTheme") ?? "classic"
        boardFlipped = UserDefaults.standard.object(forKey: "raichuBoardFlipped") as? Bool ?? true
        hapticsEnabled = UserDefaults.standard.object(forKey: "raichuHaptics") as? Bool ?? true
        HapticManager.shared.isEnabled = hapticsEnabled
    }

    func setTheme(_ name: String) {
        themeName = name
    }

    enum Tab {
        static let play = 0
        static let online = 1
        static let history = 2
        static let profile = 3
    }

    func flipBoard() {
        boardFlipped.toggle()
    }

    func enterReplay(step: Int = 0) {
        replayMode = true
        replayStep = step
    }

    func exitReplay() {
        replayMode = false
        replayStep = 0
    }

    func setReplayStep(_ step: Int) {
        replayStep = step
    }
}
